### ISSUE TITLE
Look into Tab Grouping support

### ISSUE THREAD

**Issue by mohnish** [mohnish/rearrange-tabs#37]
Chrome has introduced a new concept of Grouping Tabs. Look into any supported public APIs that can be leveraged to incorporate in the _Rearrange Tabs_ Chrome extension: https://blog.google/products/chrome/manage-tabs-with-google-chrome/

**Comment by dannypernik** [comment]
Hi @mohnish, I was very happy to see that your extension provides some keyboard accessibility for tab groups where there is none natively. I was pleasantly surprised that moving tabs left or right will move it into a closed group (and open that group). I am also happy that moving a tab to the beginning or end of the tab list does not add it to a tab group at the beginning or end of the list.

The biggest keyboard shortcut functionality I'm currently missing is the ability to navigate into a closed tab group – without rearranging the tabs. Currently, when switching from one tab to the next with the native cmd+option+left/right on mac, closed tab groups are skipped altogether.

My desired functionality would be if I'm moving into a closed group, it would open, and if I was moving out of an opened group, it would close. I imagine this being very satisfying visually – traversing from one tab group to the next while keeping only one group open at a time.

I realize you may consider this outside of the scope of the current extension, but I think you could make a lot of people very happy by adding it. It would be incredible if you could keep the current native shortcut for navigating left/right but add traversal of tab groups.

**Comment by mohnish** [comment]
Hi @dannypernik 👋 

I appreciate your input. I will consider this feature. Personally, I like the native functionality where the tab opens up the closed tab group. But, I also see how this is different UX from tab switch. I will think more about it and see if I can come up with a consistent UX for both the cases. I really like the idea of collapsing the tab group once the tab is out of it. 👍

**Comment by dannypernik** [comment]
Oops, I read this three weeks ago and thought I responded. Glad you like the idea. Regarding consistent UX whether moving a tab or just moving focus, I think it makes sense that a closed group should open when entered in both cases, and an opened focused group should close when exited in both cases.

**Comment by dannypernik** [comment]
@mohnish I've never created a Chrome extension but would love to attempt to contribute if you're willing to point me in the right direction.

**Comment by mohnish** [comment]
@dannypernik I love and appreciate your attitude. Yes, I can point you in the right direction.

* Since you've never created an extension before, I suggest getting an idea on how to do so. I found this [tutorial here that might be helpful](https://www.youtube.com/watch?v=0n809nd4Zu4)
* [Chrome's Tab API](https://developer.chrome.com/docs/extensions/reference/tabs/) that let's you handle grouping, [specifically this method](https://developer.chrome.com/docs/extensions/reference/tabs/#method-group)
* [This is where](https://github.com/mohnish/rearrange-tabs/blob/main/rearrange.js#L29-L41) we handle the commands. Introducing these commands would be done in the [manifest.json](https://github.com/mohnish/rearrange-tabs/blob/main/manifest.json#L11-L43)
* The setup itself is documented in the [README.md](https://github.com/mohnish/rearrange-tabs/blob/main/README.md#development) file

These should be good enough for you to get started on it. Hope you enjoy it. Good luck and have fun. HMU if you have any other questions. I'd definitely help you with that.

**Comment by dannypernik** [comment]
I'll see what I can do!
From [this list](https://stackoverflow.com/questions/66701875/which-chrome-shortcuts-cant-be-overwritten), it appears that the default shortcut for prev/next tab (ctrl/cmd + alt + left/right arrow) can be overwritten. Are you comfortable with adding the grouping functionality to the default keyboard shortcut?

**Comment by mohnish** [comment]
@dannypernik That'll basically break the workflows of the users and we don't want that. Pick a shortcut that's a new combination. As long as it doesn't conflict with the most common shortcuts, we should be good. You can pick one initially and see how it feels over the span of development and based on that feedback, you can always update it. You'll quickly pick up the pros and cons of a particular keyboard shortcut as you use it.

**Comment by dannypernik** [comment]
@mohnish ok. It's hard for me to imagine that users would want the default behavior that completely skips closed groups, but I know it would be an unexpected change for some. Unfortunately it seems that I will not be able to set ctrl/cmd + alt + left/right arrow as the custom shortcut for my own purposes. I'll find an alternative.

**Comment by mohnish** [comment]
@dannypernik In that case, use the same commands (for moving a tab left/right) and just update the functionality to behave the way you suggested.

**Comment by dannypernik** [comment]
@mohnish to be clear, I'm not referring to your shortcut to move tabs, rather Chrome's shortcut to move focus between tabs. The default shortcut for moving focus disregards closed tab groups, which I personally find to be egregious:

https://github.com/mohnish/rearrange-tabs/assets/3347095/2855a250-5a31-4f4c-a750-c2b4ef4ce9ef

**Comment by mohnish** [comment]
@dannypernik Oh I see. It's the tab switch you were talking about. I totally misunderstood the whole thing. I see what you mean now. I don't think the extension would be the right place to implement that functionality. The extension's purpose is to _Rearrange_ tabs. In your use case, we are just switching tabs and not really rearranging anything.

**Comment by dannypernik** [comment]
@mohnish it's true that I see Chrome's default tab switching via keyboard as the biggest opportunity for improvement - and that's because your extension rearranges tabs so well ;) that said, I'd love for a tab group to collapse when a tab is moved out of it using your shortcut. I'm still interested in seeing if I can make this happen.

Furthermore, I believe there is a case to be made for adding tab group keyboard accessibility to this extension. You have already vastly improved keyboard accessibility of tabs. Why not offer an adjacent enhancement that people have wanted for a long time?

https://support.google.com/chrome/thread/82128623/tab-groups-are-in-need-of-keyboard-accessibility?hl=en

https://support.google.com/chrome/thread/181852353/tab-groups?hl=en

![image](https://github.com/mohnish/rearrange-tabs/assets/3347095/44dac011-6475-4b15-a3d0-37210a09d389)

I feel you!! 😄  I constantly switch tabs (and do as much as possible) with the keyboard, so I currently have to choose between tab groups and keyboard shortcuts.

**Comment by dannypernik** [comment]
![image](https://github.com/mohnish/rearrange-tabs/assets/3347095/a622a752-5a36-4ad6-8f8d-087bd5d20468)

I feel you!! 😄

**Comment by mohnish** [comment]
@dannypernik 

>I'd love for a tab group to collapse when a tab is moved out of it using your shortcut. I'm still interested in seeing if I can make this happen

💯 Go for it!

**Comment by dannypernik** [comment]
@mohnish if I figured out how to handle switching tabs with groups, would I be able to fork your extension for my own use? Am I correct to think that the MIT license means I could even offer a fork of your extension in the marketplace? I don't know if I'd pursue it, but curious to get your thoughts.

**Comment by mohnish** [comment]
@dannypernik Absolutely. You can do whatever you want with the fork. In fact, I even enocourage it. All I ask is to not use the same name. Anything else is fair game. I meant to have people use the code to build something fun by themselves and that's why I used the MIT license. Go for it.

**Comment by dannypernik** [comment]
That's wonderful and very encouraging. Thank you. I would definitely attribute Rearrange Tabs as the primary author. When I have time, I'll start with the "collapse group on exit" functionality and submit a PR, then go from there.

**Comment by mohnish** [comment]
Since it has been a while, I'm going to close this.
